### PR TITLE
DEVTOOLS-48 Added a create new API command

### PR DESCRIPTION
**Pull request by fschargorosdky**
* Added a command to create API from scratch
 * Creates the API
 * Sets the root RAML
 * Pushes all files
* Log more actions in the push to let the user know what's going on
* Changed commands interface to make them callable by other commands
 * The execute function takes the parsed params

### CODE REVIEW

File: lib/controllers/createController.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  return {
+    create: create
+  };
+
+  function create(newApi) {
+    return apiPlatformService.createAPI(newApi.bizGroup, newApi.api,
+        newApi.apiVersion, newApi.rootRamlPath)
+      .then(function (api) {
+        return workspaceRepository.get()
+          .then(function (workspace) {
```
Comment (juancoen) [inline lib/controllers/createController.js]: avoid nesting then calls this way. Store needed values in function variables or use bluebird props

File: lib/controllers/createController.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            return fileSystemRepository.getFileHash(newApi.rootRamlPath)
+              .then(function (rootRamlHash) {
+                workspace.files = [
+                  {
+                    path: path.sep + newApi.rootRamlPath,
```
Comment (juancoen) [inline lib/controllers/createController.js]: I'm not sure about using path.sep as this path needs to be compared with API Platforms paths and I think it always use '/'

File: lib/repositories/apiPlatformRepository.js [hunk 3]
```diff
@@ -17,6 +19,51 @@ module.exports = function (contextHolder, errors,
...
   };
 
+  function addRootRaml(organizationId, apiId, apiVersionId, rootRaml) {
+    var newRootRaml = {
+      isDirectory: false,
+      name: path.basename(rootRaml.path),
+      data: rootRaml.data
+    };
+    return apiClient(superagent.post(
+          apiPlatformUrl +
```
Comment (juancoen) [inline lib/repositories/apiPlatformRepository.js]: too much indentation. use 4 spaces

File: lib/services/apiPlatformService.js [hunk 4]
```diff
@@ -16,6 +17,28 @@ module.exports = function (apiPlatformRepository, fileSystemRepository,
...
   };
 
+  function createAPI(organizationId, apiName, versionName, rootRamlPath) {
+    return apiPlatformRepository
+      .createAPI(organizationId, apiName, versionName)
+      .then(function (api) {
+        logger.info(messages.apiCreated());
+        logger.info(messages.uploadingRootRaml());
+        return fileSystemRepository.getFile(rootRamlPath)
+          .then(function (rootRaml) {
```
Comment (juancoen) [inline lib/services/apiPlatformService.js]: avoid nested then calls

File: lib/services/userOrganizationService.js [hunk 5]
```diff
@@ -13,6 +13,17 @@ module.exports = function (csRepository) {
         return _.map(user.memberOfOrganizations, function (bizGroup) {
           return _.pick(bizGroup, ['id', 'name']);
         });
+      })
+      .catch(function (err) {
```
Comment (juancoen) [inline lib/services/userOrganizationService.js]: this shouldn't be here, move to csRepository.getUserInfo call

File: lib/controllers/createController.js (hunk 1 above)
Comment (fschargorosdky) [inline lib/controllers/createController.js]: Done

File: lib/controllers/createController.js (hunk 2 above)
Comment (fschargorosdky) [inline lib/controllers/createController.js]: Done

File: lib/repositories/apiPlatformRepository.js (hunk 3 above)
Comment (fschargorosdky) [inline lib/repositories/apiPlatformRepository.js]: Done

File: lib/services/apiPlatformService.js (hunk 4 above)
Comment (fschargorosdky) [inline lib/services/apiPlatformService.js]: Done

File: lib/services/userOrganizationService.js (hunk 5 above)
Comment (fschargorosdky) [inline lib/services/userOrganizationService.js]: Done

File: tests/unit/createCommand.test.js [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+      };
+      createCommand.execute(args)
+        .then(function () {
+          asserts.calledOnceWithExactly(createControllerStub.create, [args]);
+          asserts.calledOnceWithExactly(setupCommandStub.execute, [{
+            bizGroup: api.organizationId,
+            api: api.id,
+            apiVersion: api.version.id
+          }]);
+          pushCommandStub.execute.calledOnce.should.be.true();
```
Comment (juancoen) [inline tests/unit/createCommand.test.js]: assert if pushCommandStub was called with or without parameters

File: tests/unit/createController.test.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+    fileSystemRepositoryStub.getFileHash =
+      sinon.stub().returns(BPromise.resolve(fileHash));
+
+    workspaceRepositoryStub.get = sinon.stub().returns(
+      BPromise.resolve(currentWorkspace));
+    workspaceRepositoryStub.update = sinon.stub().returns(BPromise.resolve());
+  });
+
+  describe('setup', run(function (createController) {
```
Comment (juancoen) [inline tests/unit/createController.test.js]: you are testing 'create', not 'setup'

**Comment by juancoen** [comment]
add test for unknown errors in csRepository.getUserInfo

**Comment by fschargorosdky** [comment]
Added the tests for csRepository.getUserInfo: code coverage 100%

File: tests/unit/createCommand.test.js (hunk 6 above)
Comment (fschargorosdky) [inline tests/unit/createCommand.test.js]: Done

File: tests/unit/createController.test.js (hunk 7 above)
Comment (fschargorosdky) [inline tests/unit/createController.test.js]: Done

**Comment by juancoen** [comment]
looks god to me

**Pull request merged**